File: src/_03_BehavioralPattern/_03_11_Visitor/Watch.js
import Device from "./Device.js";

import Rectangle from "./Rectangle.js";
import Triangle from "./Triangle.js";
import Circle from "./Circle.js";
import Star from "./Star.js";

/**
 * @typedef { import("./Shape").default } Shape
 */

export default class Watch extends Device {
  constructor() {
    super();
  }

  /**
   * @override
   * @param { Shape } shape 
   */
  print(shape) {
    switch(shape.constructor) {
      case Rectangle: {
        console.log("= Print Rectangle To Watch");
        break;
      }

      case Triangle: {
        console.log("= Print Triangle To Watch");
        break;
      }

      case Circle: {
        console.log("= Print Circle to Watch");
        break;
      }

      case Star: {
        console.log("= Print Star To Watch");
        break;
      }
    }
  }
}

File: src/_03_BehavioralPattern/_03_11_Visitor/java/after/Watch.java
package _03_BehavioralPattern._03_11_Visitor.java.after;

public class Watch implements Device {

  @Override
  public void print(Rectangle rectangle) {
    System.out.println("= Print Rectangle To Watch");
  }

  @Override
  public void print(Circle circle) {
    System.out.println("= Print Circle To Watch");
  }

  @Override
  public void print(Triangle triangle) {
    System.out.println("= Print Triangle To Watch");
  }

  @Override
  public void print(Star start) {
    System.out.println("= Print Star To Watch");
  }
  
}
